ogr/factory.py

```diff
@@ -117,4 +117,46 @@ def get_service_class(
     for service, service_kls in mapping.items():
         if service in url:
             return service_kls
+
+    return None
+
+
+def get_service_class(
+    url: str, service_mapping_update: Dict[str, Type[GitService]] = None
+) -> Type[GitService]:
+    """
+    Get the matching service class from the url.
+
+    :param url: str (url of the project, e.g. "https://github.com/packit-service/ogr")
+    :param service_mapping_update: custom mapping from  service url (str) to service class
+    :return: Matched class (subclass of GitService)
+    """
+    service_kls = get_service_class_or_none(
+        url=url, service_mapping_update=service_mapping_update
+    )
+    if service_kls:
+        return service_kls
     raise OgrException("No matching service was found.")
+
+
+def get_instances_from_dict(instances: dict) -> Set[GitService]:
+    services = []
+    for key, value in instances.items():
+        service_kls = get_service_class_or_none(url=key)
+        if not service_kls:
+            if "type" not in value:
+                raise OgrException(f"No matching service was found for url '{key}'.")
+            service_type = value["type"]
+            if service_type not in _SERVICE_MAPPING:
+                raise OgrException(
+                    f"No matching service was found for type '{service_type}'."
+                )
+
+            service_kls = _SERVICE_MAPPING[service_type]
+            value.setdefault("instance_url", key)
+            del value["type"]
```

Comment:
I'd probably log.warn in the else branch: the user should know that the key is unknown to us

Reply:
In the case of unknown key and specified `type`?

It is not an exceptional behaviour -- you are giving another info and you know that the mapping does not work.

Reply:
I meant this:
```python
get_instances_from_dict({
  "github.com": "token": "asd",
  "trolololo.lol": "token": "qwe"
})  
```
I'd say that user should know that trololo.lol is not in our mapping. I'd personally even raise an exc. With the current implementation, you just skip over it.

Reply:
okay, now that I look at the code, I'm confused actually.

Reply:
Ok, the code needs some care/docs.

We are raising an exception when:
- do not find the key in mapping and there is no `type` in the `value` dictionary
    - (=we do not know the mapping)
- do not find the key in mapping, there is `type` in the `value` dictionary, but it is not in the mapping
    - (=we do not know the mapping, user specify custom mapping, but we do not know it)

Reply:
got it, thanks for explanation, lgtm